### PR TITLE
clipboard: close stdout when copying via xclip

### LINKED ISSUE #7958: xclip clipboard test fails with `target STRING not available`
- `nvim --version`: current head 6452831cf985bd31d16eb5a04b345038f45d203f
- Operating system/version: nixos
- Terminal name/version: termite

When trying to run the functionaltests in nixos with lua5.1, I get:
```
[ SKIPPED  ] ...cratch/neovim/test/functional/provider/ruby_spec.lua @ 18: Missing neovim RubyGem.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] .../scratch/neovim/test/functional/eval/system_spec.lua @ 330: system() with a program that doesn't close stdout will exit properly after passing input
.../scratch/neovim/test/functional/eval/system_spec.lua:332: Expected objects to be the same.
Passed in:
(string) 'Error: target STRING not available
'
Expected:
(string) 'clip-data'

stack traceback:
	(tail call): ?
	.../scratch/neovim/test/functional/eval/system_spec.lua:332: in function <.../scratch/neovim/test/functional/eval/system_spec.lua:330>

[  FAILED  ] .../scratch/neovim/test/functional/eval/system_spec.lua @ 515: systemlist() with a program that doesn't close stdout will exit properly after passing input
.../scratch/neovim/test/functional/eval/system_spec.lua:518: Expected objects to be the same.
Passed in:
(table) {
 *[1] = 'Error: target STRING not available' }
Expected:
(table) {
 *[1] = 'clip'
  [2] = 'data' }

stack traceback:
	(tail call): ?
	.../scratch/neovim/test/functional/eval/system_spec.lua:518: in function <.../scratch/neovim/test/functional/eval/system_spec.lua:515>

```

According to https://unix.stackexchange.com/questions/316715/xclip-works-differently-in-interactive-and-non-interactive-shells, it doesn't seem to be nixos specific but might happen on other platforms.

If I transpose what happens in the functionaltest to shell, it fails too:
```
sh -c 'echo "clip-data" | xclip -i -loops 1 -selection clipboard'
sh -c 'xclip -o -selection clipboard'
Error: target STRING not available
```
With `sh -i -c` flag the previous test fails too.
I wonder if the root cause an be fixed or if xclip should be replaced by xsel in the test as it seems more robust according to SO.

### PR DESCRIPTION
**Pull request by janlazo**
Close https://github.com/neovim/neovim/issues/7958

https://github.com/astrand/xclip/issues/20

### CODE REVIEW

**Comment by blueyed** [comment]
Looks like a good idea, but also check https://github.com/neovim/neovim/issues/7958#issuecomment-502498751 - I can check if this fixes the issue I was seeing there also.

**Comment by janlazo** [comment]
I tested this with parcellite running as my clipboard manager on Ubuntu Xenial.Copied text to clipboard via Vim and then ran `make oldtest`.

**Comment by janlazo** [comment]
Seems to have no effect if there is no clipboard manager. I'll try a different approach.

**Comment by janlazo** [comment]
I give up. Clearing the clipboard before tests in runnvim.sh has no effect.

If there is a clipboard manager running, prefer xclip. If there is no clipboard manager running, prefer xsel and ignore xclip and similar clipboard binaries with same error. How can this be implemented?

**Comment by justinmk** [comment]
@janlazo it sounds like this is strictly an improvement though, correct? Even if it's not a total solution.

Would be worth a comment in the code and commit message, though.

**Comment by janlazo** [comment]
Is there something else missing with the updated commit?

**Comment by justinmk** [comment]
> Is there something else missing with the updated commit?

No, somehow I didn't get notified. Nice commit message!

File: runtime/autoload/provider/clipboard.vim [hunk 1]
```diff
@@ -172,6 +172,11 @@ function! s:clipboard.set(lines, regtype, reg) abort
   if jobid > 0
     call jobsend(jobid, a:lines)
     call jobclose(jobid, 'stdin')
+    " xclip does not close stdout,stderr when receiving input via stdin
+    if argv[0] ==# 'xclip'
+      call jobclose(jobid, 'stdout')
+      call jobclose(jobid, 'stderr')
```
Comment (bfredl) [inline runtime/autoload/provider/clipboard.vim]: I don't think stderr should have been closed here.

Comment (janlazo) [reply runtime/autoload/provider/clipboard.vim]: Closing stdout is insufficient.

Comment (janlazo) [reply runtime/autoload/provider/clipboard.vim]: `Test_display_registers` fails again with `target STRING not available` error.

Comment (janlazo) [reply runtime/autoload/provider/clipboard.vim]: https://github.com/neovim/neovim/pull/11661 Take 2.

**Pull request merged**